### ISSUE TITLE
Proposal: Request/Response interceptors

### ISSUE THREAD

**Issue by amckinney** [twitchtv/twirp#265]
In gRPC, interceptors are able to access the request and response objects from
the client and/or server.

Twirp supports similar middleware functionality via its `ClientHooks` and
`ServerHooks` constructs, but it's not currently possible to inspect the
request or response objects like gRPC.

This proposes a new `Interceptors` option for both the `ClientHooks` and
`ServerHooks` types.

For `ClientHooks` the user-provided interceptors would be able to take action
 * After the request structure is marshaled
 * Before the request is sent over the wire
 * After the response is returned

 Borrowing from Twirp's generated code, in `doProtobufRequest`,
 this would look something along the lines of the following:

 ```go
 func doProtoRequest(...) (context.Context, error) {
  req, err := newRequest(ctx, url, reqBody, "application/protobuf")
  if err != nil {
    return ctx, wrapInternal(err, "could not build request")
  }
  ctx, err = callClientRequestPrepared(ctx, hooks, req)
  if err != nil {
    return ctx, err
  }

  req = req.WithContext(ctx)
  method := twirp.ChainInterceptors(hooks.Interceptors...)(doProtobufMethod(out))
  if _, err := method(ctx, req); err != nil {
    return ctx, wrapInternal(err, "oops")
  }

  return ctx, nil
}

// doProtobufMethod returns an twirp.Method that performs the client call.
// This method should always occur last in the interceptor chain.
func doProtobufMethod(out proto.Message) twirp.Method {
  return func(ctx context.Context, req interface{}) (interface{}, error) {
    resp, err := client.Do(req)
    if err != nil {
      return ctx, wrapInternal(err, "failed to do request")
    }

    defer func() {
      cerr := resp.Body.Close()
      if err == nil && cerr != nil {
        err = wrapInternal(cerr, "failed to close response body")
      }
    }()

    if err = ctx.Err(); err != nil {
      return ctx, wrapInternal(err, "aborted because context was done")
    }

    if resp.StatusCode != 200 {
      return ctx, errorFromResponse(resp)
    }

    respBodyBytes, err := ioutil.ReadAll(resp.Body)
    if err != nil {
      return ctx, wrapInternal(err, "failed to read response body")
    }
    if err = ctx.Err(); err != nil {
      return ctx, wrapInternal(err, "aborted because context was done")
    }

    if err = proto.Unmarshal(respBodyBytes, out); err != nil {
      return ctx, wrapInternal(err, "failed to unmarshal proto response")
    }

    return out, nil
  }
}
```

Then in the core `twirp` package, there would be something along the
lines of the following:

```go
type Method func(ctx context.Context, request interface{}) (interface{}, error)

type Interceptor func(Method) Method

func ChainInterceptors(interceptors ...Interceptor) Interceptor {
  // Compose the chain of interceptors into a single one.
  // ...
}
```

And we could write interceptors like the following:
```go
// Logger returns a new twirp.Interceptor for logging.
func Logger(l *zap.Logger) twirp.Interceptor {
  return func(next twirp.Method) twirp.Method {
    return func(ctx context.Context, req interface{}) (interface{}, error) {
      // Log the request ...
      resp, err := next(ctx, req) // Calls the next method in the chain, e.g. doProtobufMethod above.
      if err != nil {
        // Log the error ...
        return nil, err
      }
      // Log the response ...
      return resp, nil
    }
  }
}
```

Finally, the `twirp.ClientHooks` would need to add something like the following.
```go
type ClientHooks struct {
  // Interceptors are called after the request structure is serialized according to the associated encoding.
  Interceptors []Interceptor
}
```

Note that this is example code meant to demonstrate the proposal and is not meant
to be the recommended way of implementation. There might be unforeseen bugs and issues
with this approach, for example.

**Comment by mfridman** [comment]
Its not a bad idea, just wondering if the added complexity is worth it? Chances are you're already mounting services on a mux, why is this better handled within twirp as an interceptor as apposed to middleware earlier in the chain ?

In the provided example you could inject the logger further into the ctx through middleware and then retrieve it in the hooks (or the handler) via ctx. Likewise you could do the same for request body, etc. etc.

The nice thing about twirp is that its simple, plays nicely with http middlewares and doesn't strive to have all the added complexity as gRPC.

**Comment by amckinney** [comment]
@mfridman This isn't a bad idea, but you still aren't able to inspect the Protobuf requests and responses at this layer. You have to move the interceptor further down the chain in order to interact with the marshaled request and response.

Another compelling example of this is with an integration with [protoc-gen-validate](https://github.com/envoyproxy/protoc-gen-validate). With the combination of these features, we could write a Twirp interceptor that would validate all of our Protobuf messages that implement the `validator` interface (defined below for convenience).

```go
// validator represents the a type that can be validated.
// Note this interface might be defined elsewhere in the
// protoc-gen-validate repository.
type validator interface {
  Validate() error
}

// Validator returns a new twirp.Interceptor for validating incoming requests.
func Validator() twirp.Interceptor {
  return func(next twirp.Method) twirp.Method {
    return func(ctx context.Context, req interface{}) (interface{}, error) {
      if v, ok := req.(validator); ok {
        if err := v.Validate(); err != nil {
          return nil, err
        }
      }
      return next(ctx, req)
    }
  }
}
```

**Comment by mfridman** [comment]
> You have to move the interceptor further down the chain

Yes, after thinking about it some more my initial comment was naive and based on what we've historically done.

I do actually like this proposal b/e having access to proto message and co. opens the doors for some useful (and powerful) solutions.

Traditional middlewares along the chain can only get you so far. But brining that middleware one level down "into the rpc portions" of the code would be beneficial.

**Comment by marioizquierdo** [comment]
I like this proposal. Adding some minor complexity, but it's backwards compatible and opens up better ways to integrate Twirp with other tools, which is very important.

New exported concepts: (We will need to support them moving forward)

 * `twirp.Method`: represents a Twirp service endpoint
 * `twirp.Interceptor` wraps a Method like middleware, can mutate request and response

New functionality:

 * Middleware-like interface to wrap Twirp methods, with access to the serialized request and response objects.
 * Feels more intuitive for developers used to HTTP middleware and for developers used to gRPC interceptors.
 * Allows to manipulate/mutate the request and the response in a centralized location. This may be a drawback (more on this below).

### Interceptors compared to Hooks

The validation example in the previous comment could not be done with hooks. Twirp could add a `req interface{}` to the context that could be extracted within hooks (`req := twirp.GetRequest(ctx)`) so the validation could be implemented in a server hook, but that would not be ideal; server and client hooks would work a little different, and that would allow any hook to potentially mutate the request, which makes debugging harder.

Twirp hooks have been designed to be inspect-only, to reduce the surface where things can go wrong (e.g. when debugging a Twirp service that has a modified/broken response, you don't need to inspect hooks because you know they cannot be the problem).

This doesn't mean that interceptors are a bad idea, but we should keep them separate from hooks. Instead, we can allow them as options on the client and service constructors (e.g. on the client constructor: `twirp.WithClientInterceptor(i twirp.Interceptor)`). When something goes wrong a Twirp service that has a modified/broken response, now we know where to look, the problem may be in the interceptors, but not in the hooks.

### Naming/Design

From the naming/design point of view, if we take a closer look at [gRPC interceptors](https://medium.com/@shijuvar/writing-grpc-interceptors-in-go-bf3e7671fe48), they seem to be designed with a different interface. The interceptor types are generated to include the server info, for example: 

```go
type UnaryServerInterceptor func(ctx context.Context, req interface{}, info *UnaryServerInfo, handler UnaryHandler) (resp interface{}, err error)
```

This makes me wonder if we should rename `twirp.Interceptor` to `twirp.Middleware`, because the pattern that is being proposed here looks a lot more like HTTP middleware in Go (and has the same properties, only at the Twirp serialization level). In terms of unary vs streaming we don't need to worry on Twirp because there's only unary rpc calls (if streaming happens in the future, we can always make a new type of interceptors for that). The `info` seems convenient, it is nice to get a type-safe option for that, but I think we can use the `context` passed to the `twirp.Method` function to extract info if needed, we already expose the Method, Service and Package names, we could add more info in the future if needed.

**Comment by bufdev** [comment]
FYI let us know if we can help out here - this issue is pretty much the sole blocker for us re: Twirp :-) We're happy to lend a hand with implementation

**Comment by marioizquierdo** [comment]
I'll be happy to review a PR that implements this ;)

**Comment by amckinney** [comment]
Resolved in [v7.1.0](https://github.com/twitchtv/twirp/releases/tag/v7.1.0) - thanks!
